File: 056_merge_intervals/solution.py
from typing import List

# Input: intervals = [[1,3],[2,6],[8,10],[15,18]]
# Output: [[1,6],[8,10],[15,18]]


def solve(intervals: List[List[int]]):
    intervals.sort(key=lambda x: x[0])
    result = [intervals[0]]

    for start, end in intervals[1:]:
        last_end = result[-1][1]

        # if merge is necessary:
        if start <= last_end:
            result[-1][1] = max(last_end, end)
        else:
            result.append([start, end])
    return result


print(solve([[1, 3], [2, 6], [8, 10], [15, 18]]))
print(solve([[1, 3], [8, 10], [15, 18], [2, 6]]))
print(solve([[1, 3], [2, 6], [8, 10], [15, 18], [2, 6]]))

File: 056_merge_intervals/solution.go
package main

import (
	"fmt"
	"sort"
)

func merge(intervals [][]int) [][]int {
	var result [][]int
	sort.Slice(intervals, func(i, j int) bool {
		return intervals[i][0] < intervals[j][0]
	})
	result = append(result, intervals[0])

	for _, interval := range intervals[1:] {
		start, end := interval[0], interval[1]
		lastEnd := result[len(result)-1][1]
		if start <= lastEnd {
			result[len(result)-1][1] = max(lastEnd, end)
		} else {
			result = append(result, []int{start, end})
		}
	}

	return result
}

func main() {
	nums := [][]int{{2, 6}, {8, 10}, {15, 18}, {1, 3}}

	fmt.Println(merge(nums))
}
